crates/sui-framework/sources/collectible.move

```diff
@@ -0,0 +1,446 @@
+    /// a `TreasuryCap` for `Collectible`s.
+    struct CollectionManagerCap<phantom T: store + drop> has key, store {
+        id: UID,
+        total_supply: u64,
+        max_supply: Option<u64>
+    }
+
+
+    /// Borrow `unique_number` field.
+    public fun unique_number<T: store + drop>(self: &Collectible<T>): u64 { self.unique_number }
+
+    /// Borrow `data` field.
+    public fun info<T: store + drop>(self: &Collectible<T>): &T { &self.info }
+
+    /// Implementable method giving a mutable reference to `data`.
+    public fun info_mut<T: store + drop>(_w: T, self: &mut Collectible<T>): &mut T { &mut self.info }
+
+    /// Get ID for pointing to the `Collectible`.
+    public fun id<T: store + drop>(self: &Collectible<T>): ID { object::uid_to_inner(&self.id) }
+
+    /// Get a mutable reference to UID for allowing `transfer_to_object_id` with `Collectible` type.
+    ///
+    /// `Collectible` does not have `store` ability and it means that `transfer_to_object` call can
+    /// never be used on the `Collectible`. But there's a solution to allow it having children -
+    /// by using `transfer_to_object_id` which requires a mutable reference to UID. For theses purposes
+    /// UID needs to be conditionally exposed to outer world.
+    ///
+    /// By adding implementation for this call publicly or by using it inside the module with the
+    /// implementation, it is possible to transfer objects TO `Collectible`s.
+    public fun uid_mut<T: store + drop>(_w: T, self: &mut Collectible<T>): &mut UID { &mut self.id }
+
+
+    /// Create a Collection and receive a `CollectionManagerCap`. This object
+    /// is similar to the `coin::TreasuryCap` in a way that it manages total
+    /// supply and grants the owner permission to mint and burn new `Collectible`s.
+    ///
+    /// It also holds information about the `total_supply` and prevents from
+    /// minting more than `max_supply`.
+    public fun create_collection<T: store + drop>(
+        _w: T, max_supply: Option<u64>, ctx: &mut TxContext
+    ): CollectionManagerCap<T> {
+        CollectionManagerCap {
+            max_supply,
+            id: object::new(ctx),
+            total_supply: 0,
+        }
+    }
+
+    /// Mint a new `Collectible` using `CollectionManagerCap`.
+    public fun mint<T: store + drop>(
+        cap: &mut CollectionManagerCap<T>, info: T, ctx: &mut TxContext
+    ): Collectible<T> {
+        cap.total_supply = cap.total_supply + 1;
+
+        // if `max_supply` is set -> enforce it
+        if (option::is_some(&cap.max_supply)) {
+            assert!(cap.total_supply <= *option::borrow(&cap.max_supply), EMaxSupplyReached);
+        };
+
+        Collectible {
+            info,
+            id: object::new(ctx),
+            edition: *&cap.max_supply,
+            unique_number: cap.total_supply,
+        }
+    }
+
+
+    /// Implementable `transfer::transfer` function.
+    public fun transfer<T: store + drop>(_w: T, nft: Collectible<T>, to: address) {
```

Comment:
Isn't it the case that someone can create a package with their implementation of `transfer::transfer(nft, to);` without having to instantiate the witness object?

In another words, by my understanding of the Move language, the witness argument doesn't serve any purpose here. What am I missing?

Reply:
Turns out that since `Collectible` has only `key` ability and not `store`, it cannot be transferred outside of the module which defined it.